Support clang under Visual Studio

A separate ticket as per @jkeiser's request 

On Windows 10 64bit, with:
- MSVC v142 - VS 2019 C++ x64/x86 build tools (v14.26)
- Windows 10 SDK (10.0.18362.0)
- Clang 10.0.0

In a folder named check, I put `simdjson.h`, `simdjson.cpp` and `check.cpp`:

```
#include "simdjson.h"

int main () {

    // simdjson::dom::parser parser;
    // simdjson::dom::element doc = parser.parse("[1,2,3]"_padded);

}
```

Then, from the folder in the command line, I run:


` clang++ -std=c++17 -o check.exe check.cpp simdjson.cpp`

I get:
```
simdjson.cpp:2411:20: error: use of undeclared identifier '_mm_clmulepi64_si128'
  __m128i result = _mm_clmulepi64_si128(_mm_set_epi64x(0ULL, bitmask), all_ones, 0);
                   ^
simdjson.cpp:2457:10: error: use of undeclared identifier '_blsr_u64'
  return _blsr_u64(input_num);
         ^
simdjson.cpp:2462:27: error: use of undeclared identifier '_lzcnt_u64'; did you mean '__tzcnt_u64'?
  return static_cast<int>(_lzcnt_u64(input_num));
                          ^~~~~~~~~~
                          __tzcnt_u64
C:\Program Files\LLVM\lib\clang\10.0.0\include\bmiintrin.h:87:1: note: '__tzcnt_u64' declared here
__tzcnt_u64(unsigned long long __X)
^
simdjson.cpp:2516:5: error: unknown type name '__m256i'
    __m256i value;
    ^
simdjson.cpp:2522:30: error: unknown type name '__m256i'
    really_inline base(const __m256i _value) : value(_value) {}
                             ^
simdjson.cpp:2525:34: error: unknown type name '__m256i'
    really_inline operator const __m256i&() const { return this->value; }
                                 ^
simdjson.cpp:2548:31: error: unknown type name '__m256i'
    really_inline base8(const __m256i _value) : base<simd8<T>>(_value) {}
                              ^
simdjson.cpp:2566:37: error: unknown type name '__m256i'
    really_inline simd8<bool>(const __m256i _value) : base8<bool>(_value) {}
                                    ^
simdjson.cpp:2563:66: error: use of undeclared identifier '_mm256_set1_epi8'; did you mean '_mm_set1_epi8'?
    static really_inline simd8<bool> splat(bool _value) { return _mm256_set1_epi8(-(!!_value)); }
                                                                 ^~~~~~~~~~~~~~~~
                                                                 _mm_set1_epi8
C:\Program Files\LLVM\lib\clang\10.0.0\include\emmintrin.h:3842:1: note: '_mm_set1_epi8' declared here
_mm_set1_epi8(char __b)
^
simdjson.cpp:2563:66: error: no viable conversion from returned value of type '__m128i' (vector of 2 'long long' values) to function return type 'simd8<bool>'
    static really_inline simd8<bool> splat(bool _value) { return _mm256_set1_epi8(-(!!_value)); }
                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
simdjson.cpp:2562:10: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from '__m128i' (vector of 2 'long long' values) to 'const simdjson::haswell::simd::simd8<bool> &' for 1st argument
  struct simd8<bool>: base8<bool> {
         ^
simdjson.cpp:2562:10: note: candidate constructor (the implicit move constructor) not viable: no known conversion from '__m128i' (vector of 2 'long long' values) to 'simdjson::haswell::simd::simd8<bool> &&' for 1st argument
  struct simd8<bool>: base8<bool> {
         ^
simdjson.cpp:2568:19: note: candidate constructor not viable: no known conversion from '__m128i' (vector of 2 'long long' values) to 'bool' for 1st argument
    really_inline simd8<bool>(bool _value) : base8<bool>(splat(_value)) {}
                  ^
simdjson.cpp:2565:35: error: member initializer 'base8' does not name a non-static data member or base class
    really_inline simd8<bool>() : base8() {}
                                  ^~~~~~~
simdjson.cpp:2568:46: error: type 'base8<bool>' is not a direct or virtual base of 'simdjson::haswell::simd::simd8<bool>'
    really_inline simd8<bool>(bool _value) : base8<bool>(splat(_value)) {}
                                             ^~~~~~~~~~~
simdjson.cpp:2570:51: error: use of undeclared identifier '_mm256_movemask_epi8'; did you mean '_mm_movemask_epi8'?
    really_inline int to_bitmask() const { return _mm256_movemask_epi8(*this); }
                                                  ^~~~~~~~~~~~~~~~~~~~
                                                  _mm_movemask_epi8
C:\Program Files\LLVM\lib\clang\10.0.0\include\emmintrin.h:4399:1: note: '_mm_movemask_epi8' declared here
_mm_movemask_epi8(__m128i __a)
^
simdjson.cpp:2570:72: error: no viable conversion from 'const simdjson::haswell::simd::simd8<bool>' to '__m128i' (vector of 2 'long long' values)
    really_inline int to_bitmask() const { return _mm256_movemask_epi8(*this); }
                                                                       ^~~~~
C:\Program Files\LLVM\lib\clang\10.0.0\include\emmintrin.h:4399:27: note: passing argument to parameter '__a' here
_mm_movemask_epi8(__m128i __a)
                          ^
simdjson.cpp:2571:46: error: use of undeclared identifier '_mm256_testz_si256'
    really_inline bool any() const { return !_mm256_testz_si256(*this, *this); }
                                             ^
simdjson.cpp:2572:64: error: invalid operands to binary expression ('const simdjson::haswell::simd::simd8<bool>' and 'bool')
    really_inline simd8<bool> operator~() const { return *this ^ true; }
                                                         ~~~~~ ^ ~~~~
simdjson.cpp:2596:39: error: unknown type name '__m256i'
    really_inline base8_numeric(const __m256i _value) : base8<T>(_value) {}
                                      ^
simdjson.cpp:2685:31: error: unknown type name '__m256i'
    really_inline simd8(const __m256i _value) : base8_numeric<int8_t>(_value) {}
                              ^
simdjson.cpp:2684:29: error: type 'base8_numeric<int8_t>' (aka 'base8_numeric<signed char>') is not a direct or virtual base of 'simdjson::haswell::simd::simd8<int8_t>'
    really_inline simd8() : base8_numeric<int8_t>() {}
                            ^~~~~~~~~~~~~~~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```

Hey,

Edited and reran, I get - 

```
simdjson.cpp:2412:20: error: use of undeclared identifier '_mm_clmulepi64_si128'
  __m128i result = _mm_clmulepi64_si128(_mm_set_epi64x(0ULL, bitmask), all_ones, 0);
                   ^
simdjson.cpp:2458:10: error: use of undeclared identifier '_blsr_u64'
  return _blsr_u64(input_num);
         ^
simdjson.cpp:2463:27: error: use of undeclared identifier '_lzcnt_u64'; did you mean '__tzcnt_u64'?
  return static_cast<int>(_lzcnt_u64(input_num));
                          ^~~~~~~~~~
                          __tzcnt_u64
C:\Program Files\LLVM\lib\clang\10.0.0\include\bmiintrin.h:87:1: note: '__tzcnt_u64' declared here
__tzcnt_u64(unsigned long long __X)
^
simdjson.cpp:2517:5: error: unknown type name '__m256i'
    __m256i value;
    ^
simdjson.cpp:2523:30: error: unknown type name '__m256i'
    really_inline base(const __m256i _value) : value(_value) {}
                             ^
simdjson.cpp:2526:34: error: unknown type name '__m256i'
    really_inline operator const __m256i&() const { return this->value; }
                                 ^
simdjson.cpp:2549:31: error: unknown type name '__m256i'
    really_inline base8(const __m256i _value) : base<simd8<T>>(_value) {}
                              ^
simdjson.cpp:2567:37: error: unknown type name '__m256i'
    really_inline simd8<bool>(const __m256i _value) : base8<bool>(_value) {}
                                    ^
simdjson.cpp:2564:66: error: use of undeclared identifier '_mm256_set1_epi8'; did you mean '_mm_set1_epi8'?
    static really_inline simd8<bool> splat(bool _value) { return _mm256_set1_epi8(-(!!_value)); }
                                                                 ^~~~~~~~~~~~~~~~
                                                                 _mm_set1_epi8
C:\Program Files\LLVM\lib\clang\10.0.0\include\emmintrin.h:3842:1: note: '_mm_set1_epi8' declared here
_mm_set1_epi8(char __b)
^
simdjson.cpp:2564:66: error: no viable conversion from returned value of type '__m128i'
      (vector of 2 'long long' values) to function return type 'simd8<bool>'
    static really_inline simd8<bool> splat(bool _value) { return _mm256_set1_epi8(-(!!_value)); }
                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
simdjson.cpp:2563:10: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from
      '__m128i' (vector of 2 'long long' values) to 'const simdjson::haswell::simd::simd8<bool> &' for 1st argument
  struct simd8<bool>: base8<bool> {
         ^
simdjson.cpp:2563:10: note: candidate constructor (the implicit move constructor) not viable: no known conversion from
      '__m128i' (vector of 2 'long long' values) to 'simdjson::haswell::simd::simd8<bool> &&' for 1st argument
  struct simd8<bool>: base8<bool> {
         ^
simdjson.cpp:2569:19: note: candidate constructor not viable: no known conversion from '__m128i'
      (vector of 2 'long long' values) to 'bool' for 1st argument
    really_inline simd8<bool>(bool _value) : base8<bool>(splat(_value)) {}
                  ^
simdjson.cpp:2566:35: error: member initializer 'base8' does not name a non-static data member or base class
    really_inline simd8<bool>() : base8() {}
                                  ^~~~~~~
simdjson.cpp:2569:46: error: type 'base8<bool>' is not a direct or virtual base of
      'simdjson::haswell::simd::simd8<bool>'
    really_inline simd8<bool>(bool _value) : base8<bool>(splat(_value)) {}
                                             ^~~~~~~~~~~
simdjson.cpp:2571:51: error: use of undeclared identifier '_mm256_movemask_epi8'; did you mean '_mm_movemask_epi8'?
    really_inline int to_bitmask() const { return _mm256_movemask_epi8(*this); }
                                                  ^~~~~~~~~~~~~~~~~~~~
                                                  _mm_movemask_epi8
C:\Program Files\LLVM\lib\clang\10.0.0\include\emmintrin.h:4399:1: note: '_mm_movemask_epi8' declared here
_mm_movemask_epi8(__m128i __a)
^
simdjson.cpp:2571:72: error: no viable conversion from 'const simdjson::haswell::simd::simd8<bool>' to '__m128i'
      (vector of 2 'long long' values)
    really_inline int to_bitmask() const { return _mm256_movemask_epi8(*this); }
                                                                       ^~~~~
C:\Program Files\LLVM\lib\clang\10.0.0\include\emmintrin.h:4399:27: note: passing argument to parameter '__a' here
_mm_movemask_epi8(__m128i __a)
                          ^
simdjson.cpp:2572:46: error: use of undeclared identifier '_mm256_testz_si256'
    really_inline bool any() const { return !_mm256_testz_si256(*this, *this); }
                                             ^
simdjson.cpp:2573:64: error: invalid operands to binary expression ('const simdjson::haswell::simd::simd8<bool>' and
      'bool')
    really_inline simd8<bool> operator~() const { return *this ^ true; }
                                                         ~~~~~ ^ ~~~~
simdjson.cpp:2597:39: error: unknown type name '__m256i'
    really_inline base8_numeric(const __m256i _value) : base8<T>(_value) {}
                                      ^
simdjson.cpp:2686:31: error: unknown type name '__m256i'
    really_inline simd8(const __m256i _value) : base8_numeric<int8_t>(_value) {}
                              ^
simdjson.cpp:2685:29: error: type 'base8_numeric<int8_t>' (aka 'base8_numeric<signed char>') is not a direct or virtual

      base of 'simdjson::haswell::simd::simd8<int8_t>'
    really_inline simd8() : base8_numeric<int8_t>() {}
                            ^~~~~~~~~~~~~~~~~~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
```

So why did I remember this? If you Google it, you will find a whole thread with me in it where I observe that though Visual Studio has integrated clang, the intrinsics are not yet supported. This was circa 2016.

Things have changed and clang under Windows does support the intrinsics... but not in the way we expect.



Ok. So I can very nearly build by adding just these lines... right after IS_X86_64 is defined. It is not good code in the sense that it undefines _MSC_VER which is a big nono, but as a proof of concept. The reason for these lines is that clang under visual studio only includes the intrinsics if the corresponding feature is flagged with a macro. Why it does this is beyond me, it is unique to my knowledge. But ok, so I force it...



```C++
#if defined(__clang__) && defined(_MSC_VER) && defined( IS_X86_64)
#define __AVX2__ 1
#define __AVX__ 1
#define __SSE3__ 1
#define __SSE4_1__ 1
#define __SSE4_2__ 1
#define __SSE_MATH__ 1
#define __SSE__ 1
#define __SSSE3__ 1
#define __PCLMUL__ 1
#define __BMI2__ 1
#define __BMI__ 1
#define __LZCNT__ 1
#define strcasecmp _stricmp

static inline int posix_memalign(void **p, size_t alignment, size_t size) {
#ifdef _MSC_VER
    *p = _aligned_malloc(size, alignment);
    return *p == nullptr;
#elif defined(__MINGW32__) || defined(__MINGW64__)
    *p = __mingw_aligned_malloc(size, alignment);
    return *p == nullptr;
#else
    return posix_memalign(p, alignment, size);
#endif
}
#undef _MSC_VER
#endif
```


Now, what happens after that is interesting... 

```
clang -o test.exe .\amalgamate_demo.cpp

(...)

.\simdjson.cpp:2869:52: error: always_inline function 'operator==' requires target feature 'avx2', but would be inlined into function   
      'operator()' that is compiled without support for 'avx2'
      return this->map( [&](simd8<T> a) { return a == mask; } ).to_bitmask();
                                                   ^
.\simdjson.cpp:2874:52: error: always_inline function 'operator<=' requires target feature 'avx2', but would be inlined into function   
      'operator()' that is compiled without support for 'avx2'
      return this->map( [&](simd8<T> a) { return a <= mask; } ).to_bitmask();
                                                   ^
.\simdjson.cpp:4406:52: error: always_inline function 'operator==' requires target feature 'sse4.2', but would be inlined into function 
      'operator()' that is compiled without support for 'sse4.2'
      return this->map( [&](simd8<T> a) { return a == mask; } ).to_bitmask();
                                                   ^
.\simdjson.cpp:4411:52: error: always_inline function 'operator<=' requires target feature 'sse4.2', but would be inlined into function 
      'operator()' that is compiled without support for 'sse4.2'
      return this->map( [&](simd8<T> a) { return a <= mask; } ).to_bitmask();
                                                   ^
3 warnings and 4 errors generated.
```

So what seems to be happening is that there is a bad interaction between the lambdas and clang under Visual Studio, specifically, under clang 10. I am pinging @jkeiser on this one.


Ok. Fine. What about using the clang that comes with Visual Studio? Then it works...

```
 C:\"Program Files (x86)"\"Microsoft Visual Studio"\2019\Community\VC\Tools\Llvm\bin\clang.exe -o test.exe .\amalgamate_demo.cpp 
```

--- a/include/simdjson.h
+++ b/include/simdjson.h
@@ -8,6 +8,12 @@
  */
 
 #include "simdjson/compiler_check.h"
+#include "simdjson/common_defs.h"
+
+SIMDJSON_PUSH_DISABLE_WARNINGS
+#if defined(_MSC_VER) && defined(__clang__)
+SIMDJSON_DISABLE_GCC_WARNING(-Wmicrosoft-include)
+#endif
 
 // Public API
 #include "simdjson/simdjson_version.h"
@@ -28,5 +34,7 @@
 #include "simdjson/inline/error.h"
 #include "simdjson/inline/padded_string.h"
 #include "simdjson/inline/parsedjson_iterator.h"
+
+SIMDJSON_POP_DISABLE_WARNINGS
 
 #endif // SIMDJSON_H

--- a/include/simdjson/common_defs.h
+++ b/include/simdjson/common_defs.h
@@ -59,7 +59,7 @@ constexpr size_t DEFAULT_MAX_DEPTH = 1024;
 
 #define ISALIGNED_N(ptr, n) (((uintptr_t)(ptr) & ((n)-1)) == 0)
 
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && !defined(__clang__)
 
   #define really_inline __forceinline
   #define never_inline __declspec(noinline)
@@ -119,7 +119,15 @@ constexpr size_t DEFAULT_MAX_DEPTH = 1024;
   #define SIMDJSON_DISABLE_DEPRECATED_WARNING SIMDJSON_DISABLE_GCC_WARNING(-Wdeprecated-declarations)
   #define SIMDJSON_POP_DISABLE_WARNINGS _Pragma("GCC diagnostic pop")
 
-  #define SIMDJSON_DLLIMPORTEXPORT
+  #if defined(_MSC_VER) && defined(__clang__)
+    #if SIMDJSON_USING_LIBRARY
+    #define SIMDJSON_DLLIMPORTEXPORT __declspec(dllimport)
+    #else
+    #define SIMDJSON_DLLIMPORTEXPORT __declspec(dllexport)
+    #endif
+  #else
+    #define SIMDJSON_DLLIMPORTEXPORT
+  #endif
 
 #endif // MSC_VER
 

--- a/src/haswell/intrinsics.h
+++ b/src/haswell/intrinsics.h
@@ -4,7 +4,14 @@
 #include "simdjson.h"
 
 #ifdef _MSC_VER
-#include <intrin.h> // visual studio
+// visual studio (or clang-cl)
+#include <intrin.h>
+#ifdef __clang__
+#include <bmiintrin.h>
+#include <lzcntintrin.h>
+#include <immintrin.h>
+#include <wmmintrin.h>
+#endif
 #else
 #include <x86intrin.h> // elsewhere
 #endif // _MSC_VER

--- a/src/simdjson.cpp
+++ b/src/simdjson.cpp
@@ -1,5 +1,9 @@
 #include "simdjson.h"
 
+#if defined(_MSC_VER) && defined(__clang__)
+SIMDJSON_DISABLE_GCC_WARNING(-Wmicrosoft-include)
+#endif
+
 /* used for http://dmalloc.com/ Dmalloc - Debug Malloc Library */
 #ifdef DMALLOC
 #include "dmalloc.h"

